File: problems/CSES/apple_division/apple_division.py
import sys
input = sys.stdin.readline

def main():
    n = int(input())
    list_p = list(map(int, input().split()))

    total = sum(list_p)
    res = float('inf')

    def dfs(i, current_sum):
        nonlocal res

        if i == n:
            res = min(res, abs(total - 2 * current_sum))
            return

        dfs(i + 1, current_sum + list_p[i])

        dfs(i + 1, current_sum)

    dfs(0, 0)

    print(res)

main()

File: problems/CSES/apple_division/apple_division.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n = 1;
    cin >> n;

    vector<ll> list_a(n);
    for (int i = 0; i < n; i++) {
        cin >> list_a[i];
    }

    ll total = accumulate(list_a.begin(), list_a.end(), 0LL);
    ll res = numeric_limits<ll>::max(); 
    
    auto dfs = [&](auto& self, int i, ll curr_sum) -> void {
        if (i == n) {
            res = min(res, abs(total - 2 * curr_sum));
            return;
        }
        self(self, i + 1, curr_sum + list_a[i]);
        self(self, i + 1, curr_sum);
    };

    dfs(dfs, 0, 0);

    cout << res << '\n';

    return 0;
}
